internal/oci/notation.go

```diff
@@ -0,0 +1,225 @@
+	"github.com/google/go-containerregistry/pkg/name"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+	"github.com/notaryproject/notation-go/dir"
+	verifier "github.com/notaryproject/notation-go/verifier"
+	"github.com/notaryproject/notation-go/verifier/truststore"
+
+	_ "github.com/notaryproject/notation-core-go/signature/cose"
+	_ "github.com/notaryproject/notation-core-go/signature/jws"
+	"github.com/notaryproject/notation-go"
+	"github.com/notaryproject/notation-go/registry"
+	"github.com/notaryproject/notation-go/verifier/trustpolicy"
+	oras "oras.land/oras-go/v2/registry/remote"
+	oauth "oras.land/oras-go/v2/registry/remote/auth"
+)
+
+// notationOptions is a struct that holds options for notation verifier
+type notationOptions struct {
+	PublicKey  []byte
+	TrustStore *trustpolicy.Document
+	Keychain   authn.Keychain
+	ROpt       []remote.Option
+	Insecure   bool
+}
+
+// NotationOptions is a function that configures the options applied to a notation verifier
+type NotationOptions func(opts *notationOptions)
+
+// WithInsecureRegistry sets notation to verify against insecure registry.
+func WithInsecureRegistry(insecure bool) NotationOptions {
+	return func(opts *notationOptions) {
+		opts.Insecure = insecure
+	}
+}
+
+// WithTrustStore sets the trust store configuration.
+func WithTrustStore(trustStore *trustpolicy.Document) NotationOptions {
+	return func(opts *notationOptions) {
+		opts.TrustStore = trustStore
+	}
+}
+
+// WithNotaryPublicKey sets the public key.
+func WithNotaryPublicKey(publicKey []byte) NotationOptions {
+	return func(opts *notationOptions) {
+		opts.PublicKey = publicKey
+	}
+}
+
+// WithNotaryRemoteOptions is a functional option for overriding the default
+// remote options used by the verifier
+func WithNotaryRemoteOptions(opts ...remote.Option) NotationOptions {
+	return func(o *notationOptions) {
+		o.ROpt = opts
+	}
+}
+
+// WithNotaryKeychain is a functional option for overriding the default
+// remote options used by the verifier
+func WithNotaryKeychain(key authn.Keychain) NotationOptions {
+	return func(o *notationOptions) {
+		o.Keychain = key
+	}
+}
+
+// NotatryVerifier is a struct which is responsible for executing verification logic
+type NotaryVerifier struct {
+	auth     authn.Keychain
+	verifier *notation.Verifier
+	opts     []remote.Option
+	insecure bool
+}
+
+// NewNotaryVerifier initializes a new NotaryVerifier
+func NewNotaryVerifier(opts ...NotationOptions) (*NotaryVerifier, error) {
+	o := notationOptions{}
+	for _, opt := range opts {
+		opt(&o)
+	}
+
+	for _, pol := range o.TrustStore.TrustPolicies {
+		certName := pol.Name
+		for _, store := range pol.TrustStores {
+			s := strings.Split(store, ":")
+			if len(s) != 2 {
+				return nil, fmt.Errorf("trust store '%s' is invalid. Trust store name must contain a store type and a store name separated by ':'. For example 'ca:fluxcd.io'", store)
+			}
+			generateTrustStore(s[0], s[1], certName, o.PublicKey)
```

Comment:
we should handle the directory cleanUp. We don't want to leave those in the source-controller /tmp/.

Reply:
instead of having to write certificates to disk and then worrying about cleanup, can we do everything in memory? i'm not super familiar with notiation, but it looks like it should be possible by writing a custom implementation of this interface that reads certificates from memory
https://github.com/notaryproject/notation-go/blob/966c6b7e42e463522cddc673bc753f67fe52de77/verifier/truststore/truststore.go#L49

Reply:
Great suggestion, @aryan9600! I've implemented that in [d840cc4](https://github.com/fluxcd/source-controller/pull/1075/commits/d840cc498aef771ca604d99b8f5b2a6c27150b41).